### PR TITLE
(Re-open #22) Refactored Warehouse ROS for pluginlib

### PR DESCRIPTION
**Pull request by 130s**
Re-opening https://github.com/ros-planning/warehouse_ros/pull/22 because it:
- is conflicted
- contains some undesired removal (changelog, .travis.yml)
- and yes, travis wasn't triggered properly.

### CODE REVIEW

File: include/warehouse_ros/database_loader.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+{
+
+class DBConnectionStub : public DatabaseConnection {
+public:
+  bool setParams(const std::string& host, unsigned port, float timeout) {return false;}
+  bool setTimeout(float timeout) {return false;}
+  bool connect() {return false;}
+  bool isConnected() {return false;}
+  void dropDatabase(const std::string& db_name) {throw warehouse_ros::DbConnectException("Database is stub");}
+  std::string messageType(const std::string& db_name, const std::string& collection_name) {throw warehouse_ros::DbConnectException("Database is stub");}
```
Comment (davetcoleman) [inline include/warehouse_ros/database_loader.h]: roscpp style guideline: "Braces, both open and close, go on their own lines (no "cuddled braces")."

**Comment by davetcoleman** [comment]
+1 I scanned all the code, though it was a lot to review thoroughly

**Comment by 130s** [comment]
Just a note for the future; even this PR seems to fail with Kinetic on Xenial, as the [test on Travis failed](https://travis-ci.org/ros-planning/warehouse_ros/jobs/137983071#L3673-L3696) and I'm seeing the same error locally on my Xenial.

```
[warehouse_ros:make] /usr/bin/c++   -DROSCONSOLE_BACKEND_LOG4CXX -DROS_PACKAGE_NAME=\"warehouse_ros\" -Dwarehouse_ros_EXPORTS -I/root/ros/ws_ci_src/src/ci_src/include -I/opt/ros/kinetic/include  -fPIC   -o CMakeFiles/warehouse_ros.dir/src/transform_collection.cpp.o -c /root/ros/ws_ci_src/src/ci_src/src/transform_collection.cpp
[warehouse_ros:make] In file included from /root/ros/ws_ci_src/src/ci_src/src/database_loader.cpp:37:0:
[warehouse_ros:make] /root/ros/ws_ci_src/src/ci_src/include/warehouse_ros/database_loader.h:78:10: error: 'scoped_ptr' in namespace 'boost' does not name a template type
[warehouse_ros:make]    boost::scoped_ptr<pluginlib::ClassLoader<warehouse_ros::DatabaseConnection> > db_plugin_loader_;
[warehouse_ros:make]           ^
[warehouse_ros:make] /root/ros/ws_ci_src/src/ci_src/src/database_loader.cpp: In member function 'void warehouse_ros::DatabaseLoader::initialize()':
[warehouse_ros:make] /root/ros/ws_ci_src/src/ci_src/src/database_loader.cpp:59:5: error: 'db_plugin_loader_' was not declared in this scope
[warehouse_ros:make]      db_plugin_loader_.reset(new pluginlib::ClassLoader<DatabaseConnection>("warehouse_ros", "warehouse_ros::DatabaseConnection"));
[warehouse_ros:make]      ^
[warehouse_ros:make] /root/ros/ws_ci_src/src/ci_src/src/database_loader.cpp: In member function 'warehouse_ros::DatabaseConnection::Ptr warehouse_ros::DatabaseLoader::loadDatabase()':
[warehouse_ros:make] /root/ros/ws_ci_src/src/ci_src/src/database_loader.cpp:69:8: error: 'db_plugin_loader_' was not declared in this scope
[warehouse_ros:make]    if (!db_plugin_loader_)
[warehouse_ros:make]         ^
[warehouse_ros:make] /root/ros/ws_ci_src/src/ci_src/src/database_loader.cpp:89:14: error: 'db_plugin_loader_' was not declared in this scope
[warehouse_ros:make]      db.reset(db_plugin_loader_->createUnmanagedInstance(db_plugin));
[warehouse_ros:make]               ^
[warehouse_ros:make] CMakeFiles/warehouse_ros.dir/build.make:62: recipe for target 'CMakeFiles/warehouse_ros.dir/src/database_loader.cpp.o' failed
[warehouse_ros:make] make[2]: *** [CMakeFiles/warehouse_ros.dir/src/database_loader.cpp.o] Error 1
[warehouse_ros:make] make[2]: *** Waiting for unfinished jobs....
[warehouse_ros:make] make[2]: Leaving directory '/root/ros/ws_ci_src/build/warehouse_ros'
[warehouse_ros:make] CMakeFiles/Makefile2:488: recipe for target 'CMakeFiles/warehouse_ros.dir/all' failed
[warehouse_ros:make] make[1]: Leaving directory '/root/ros/ws_ci_src/build/warehouse_ros'
[warehouse_ros:make] make[1]: *** [CMakeFiles/warehouse_ros.dir/all] Error 2
[warehouse_ros:make] Makefile:138: recipe for target 'all' failed
[warehouse_ros:make] make: *** [all] Error 2
Failed <<< warehouse_ros                        [ 9.1 seconds ]
```

**Comment by mikeferguson** [comment]
@130s -- is it just a missing header include? I don't see any boost includes in that header -- and I know some of the lower level stuff might have moved towards using C++11 instead of boost possibly dropping the previously included header

**Comment by 130s** [comment]
@davetcoleman Code style should be now adjusted for all files in `include` and `src`, although I just used Eclipse code formatter http://wiki.ros.org/IDEs#Auto_Formatting and don't know how reliable that is.

**Comment by 130s** [comment]
@mikeferguson thanks for the catch for the headers. I added a simple change and it builds on my Xenial host. Now waiting for Travis result to see if it builds for both Jade and Kinetic.

**Comment by 130s** [comment]
Now Travis is happy for Jade and Kinetic. 
I leave merge decision to others since this is related to a long-lasting https://github.com/ros-planning/warehouse_ros/issues/20

(So at the moment we can technically release `warehouse_ros` into J and K from `jade-devel` branch along with the changes in this PR?)

**Comment by davetcoleman** [comment]
@mikeferguson can you do the honors?

**Comment by mikeferguson** [comment]
I'll push a release later tonight

**Comment by 130s** [comment]
@mikeferguson if it's just a bloom without anything special then I can do that early afternoon. Let me know.

**Pull request merged**